File: app/src/main/java/fr/smarthomeworld/wealth/ui/Format.kt
package fr.smarthomeworld.wealth.ui

import java.text.NumberFormat
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Currency
import java.util.Locale

/** Numbers as the phone's language writes them, in the money's own unit. */
object Fmt {

    fun money(value: Double?, currency: String?, decimals: Boolean = false): String {
        if (value == null) return "—"
        val f = NumberFormat.getCurrencyInstance(Locale.getDefault())
        runCatching { f.currency = Currency.getInstance((currency ?: "EUR").uppercase()) }
        f.maximumFractionDigits = if (decimals) 2 else 0
        f.minimumFractionDigits = if (decimals) 2 else 0
        return f.format(value)
    }

    fun signedMoney(value: Double?, currency: String?): String {
        if (value == null) return "—"
        val body = money(kotlin.math.abs(value), currency, decimals = kotlin.math.abs(value) < 1000)
        return (if (value >= 0) "+" else "−") + body
    }

    fun percent(fraction: Double?): String {
        if (fraction == null) return "—"
        val v = fraction * 100
        return String.format(Locale.getDefault(), "%+.2f %%", v)
    }

    fun day(iso: String?): String {
        if (iso.isNullOrBlank()) return ""
        return runCatching {
            LocalDate.parse(iso.take(10)).format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM))
        }.getOrDefault(iso)
    }

    /** "just now", "12 min ago", "3 h ago", or the day. */
    fun since(millis: Long): String {
        if (millis <= 0) return ""
        val d = Duration.between(Instant.ofEpochMilli(millis), Instant.now())
        return when {
            d.toMinutes() < 1 -> "just now"
            d.toHours() < 1 -> "${d.toMinutes()} min ago"
            d.toDays() < 1 -> "${d.toHours()} h ago"
            else -> day(LocalDate.now().minusDays(d.toDays()).toString())
        }
    }
}

File: ios/Shared/Format.swift
import SwiftUI

/// Gain and loss, the same two colours as the Android app (contract rule 7).
extension Color {
    static let gain = Color(red: 52.0 / 255, green: 211.0 / 255, blue: 153.0 / 255)   // #34D399
    static let loss = Color(red: 248.0 / 255, green: 113.0 / 255, blue: 113.0 / 255)  // #F87171

    static func sign(_ value: Double?) -> Color {
        guard let value, value != 0 else { return .secondary }
        return value > 0 ? .gain : .loss
    }
}

/// Numbers as the phone's language writes them, in the money's own unit
/// (contract rule 3): two decimals under 1 000, none above.
enum Fmt {
    static func money(_ value: Double?, _ currency: String?, locale: Locale = .current) -> String {
        guard let value else { return "—" }
        let f = NumberFormatter()
        f.locale = locale
        f.numberStyle = .currency
        f.currencyCode = (currency ?? "EUR").uppercased()
        let decimals = abs(value) < 1000 ? 2 : 0
        f.minimumFractionDigits = decimals
        f.maximumFractionDigits = decimals
        return f.string(from: NSNumber(value: value)) ?? "\(value)"
    }

    /// A unit price: always with its decimals, whatever its size — a
    /// share at 1 234,56 € is not "1 235 €".
    static func price(_ value: Double?, _ currency: String?, locale: Locale = .current) -> String {
        guard let value else { return "—" }
        let f = NumberFormatter()
        f.locale = locale
        f.numberStyle = .currency
        f.currencyCode = (currency ?? "EUR").uppercased()
        f.minimumFractionDigits = 2
        f.maximumFractionDigits = abs(value) < 10 ? 4 : 2
        return f.string(from: NSNumber(value: value)) ?? "\(value)"
    }

    /// A number of shares: whole where it is whole, up to four places
    /// where a savings plan has made it fractional.
    static func quantity(_ value: Double?, locale: Locale = .current) -> String {
        guard let value else { return "—" }
        let f = NumberFormatter()
        f.locale = locale
        f.numberStyle = .decimal
        f.minimumFractionDigits = 0
        f.maximumFractionDigits = 4
        return f.string(from: NSNumber(value: value)) ?? "\(value)"
    }

    static func signedMoney(_ value: Double?, _ currency: String?, locale: Locale = .current) -> String {
        guard let value else { return "—" }
        return (value >= 0 ? "+" : "−") + money(abs(value), currency, locale: locale)
    }

    static func percent(_ fraction: Double?, locale: Locale = .current) -> String {
        guard let fraction else { return "—" }
        let f = NumberFormatter()
        f.locale = locale
        f.numberStyle = .decimal
        f.minimumFractionDigits = 2
        f.maximumFractionDigits = 2
        f.positivePrefix = "+"
        return (f.string(from: NSNumber(value: fraction * 100)) ?? "\(fraction * 100)") + " %"
    }

    /// "2026-09-24" or a full timestamp, as a medium date.
    static func day(_ iso: String?, locale: Locale = .current) -> String {
        guard let iso, iso.count >= 10 else { return iso ?? "" }
        let parse = DateFormatter()
        parse.locale = Locale(identifier: "en_US_POSIX")
        parse.timeZone = TimeZone(identifier: "UTC")
        parse.dateFormat = "yyyy-MM-dd"
        guard let date = parse.date(from: String(iso.prefix(10))) else { return iso }
        let show = DateFormatter()
        show.locale = locale
        show.timeZone = TimeZone(identifier: "UTC")
        show.dateStyle = .medium
        show.timeStyle = .none
        return show.string(from: date)
    }

    /// "just now", "12 min ago", "3 h ago", or the day.
    static func since(_ date: Date?, now: Date = Date()) -> String {
        guard let date else { return "" }
        let seconds = now.timeIntervalSince(date)
        if seconds < 60 { return String(localized: "just now") }
        if seconds < 86_400 {
            let f = RelativeDateTimeFormatter()
            f.unitsStyle = .abbreviated
            return f.localizedString(for: date, relativeTo: now)
        }
        return date.formatted(date: .abbreviated, time: .omitted)
    }
}

extension Color {
    /// "#7c3aed" as the dashboard writes it; anything else is nil.
    init?(hex: String?) {
        guard var s = hex?.trimmingCharacters(in: .whitespaces) else { return nil }
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6, let v = UInt32(s, radix: 16) else { return nil }
        self.init(red: Double((v >> 16) & 0xFF) / 255,
                  green: Double((v >> 8) & 0xFF) / 255,
                  blue: Double(v & 0xFF) / 255)
    }
}
